### PR TITLE
feat: add `axis=None` specialization for `Min` reducer + small refactoring for existing specializatons

### PR DESCRIPTION
**Pull request by ikrommyd**
This adds `axis=None` fast path specialization for `Min`.
We also do some cleanup at the same time and remove the existing sum/max specializations for jax.
(I am sneaking in one more deprecation warning here for jax so that it's also raised when one does `ak.jax.register_and_check()`).

### CODE REVIEW

**Comment by ikrommyd** [comment]
@ianna  if you could please check that the same speedup is there for `max` and that I didn't ruin anything, that'd be great. And that also there is a speedup for min too now.

File: src/awkward/_do.py [hunk 1]
```diff
@@ -314,14 +314,6 @@ def reduce(
...
             # so we treat them as equivalent and recurse to the axis=None specialization
             if depth == negaxis == 1:
-                # Use axis=None specialization if available
-                if (
-                    layout.backend.nplike.known_data
-                    and (specialization := original_reducer.axis_none_reducer())
-                    is not None
-                ):
-                    original_reducer = specialization
-
```
Comment (ikrommyd) [inline src/awkward/_do.py:324]: not needed as covered automatically by the recursion right below

File: src/awkward/_kernels.py [hunk 2]
```diff
@@ -70,22 +70,17 @@ def __call__(self, *args) -> Any: ...
 
 
 class NumpyKernel(BaseKernel):
-    def __init__(self, impl: Callable[..., Any], key: KernelKeyType):
-        super().__init__(impl, key)
-        # Pre-calculate which arguments are pointers.
-        # This moves the issubclass() calls out of the hot path.
-        self._is_ptr = [issubclass(t, ctypes._Pointer) for t in self._impl.argtypes]
-        self._argtypes = self._impl.argtypes
-
```
Comment (ikrommyd) [inline src/awkward/_kernels.py:79]: I did not understand this change from the PR that introduced the max specialization. These attributes are not used anywhere.

Comment (ianna) [reply src/awkward/_kernels.py:79]: I added a comment to explain it:
```python
        # Pre-calculate which arguments are pointers.
        # This moves the issubclass() calls out of the hot path.
```

File: src/awkward/_do.py (hunk 1 above)
Comment (ianna) [reply src/awkward/_do.py:324]: nope, the code path doesn't end up at the specialized reducer

File: src/awkward/_kernels.py (hunk 2 above)
Comment (ikrommyd) [reply src/awkward/_kernels.py:79]: `_is_ptr` and `_argtypes` are not used at all in `NumpyKernel` though when called. It's dead code that's never used atm.

File: src/awkward/_reducers.py [hunk 3]
```diff
@@ -718,7 +717,10 @@ def __init__(self, initial: float | None):
     def initial(self) -> float | None:
         return self._initial
 
-    def _identity_for(self, dtype: DTypeLike | None) -> float:
+    def axis_none_reducer(self) -> AxisNoneMin:
+        return AxisNoneMin(self._initial)
+
+    def _identity_for(self, dtype: DTypeLike | None) -> np.generic:
```
Comment (ianna) [inline src/awkward/_reducers.py]: note the `preferred_dtype`: Final = np.float64

File: src/awkward/_reducers.py [hunk 4]
```diff
@@ -833,10 +867,10 @@ def __init__(self, initial):
     def initial(self):
         return self._initial
 
-    def axis_none_reducer(self):
+    def axis_none_reducer(self) -> AxisNoneMax:
         return AxisNoneMax(self._initial)
 
-    def _identity_for(self, dtype: DTypeLike | None):
+    def _identity_for(self, dtype: DTypeLike | None) -> np.generic:
```
Comment (ianna) [inline src/awkward/_reducers.py]: see my comment above

**Review by ianna — CHANGES_REQUESTED** [review]
@ikrommyd - please keep the JAX code in - we need it to run the test in preparation for the next week blueprint. As for the other changes - they are not random and have been tested on the previous reducer and did contribute to the performance gains. Thanks.

File: src/awkward/_do.py (hunk 1 above)
Comment (ikrommyd) [reply src/awkward/_do.py:324]: It recursively calls `reduce` with `axis=None` one line below. Can also be seen via breakpointing.

Comment (ikrommyd) [reply src/awkward/_do.py:324]:
```
                return reduce(
                    layout=layout,
                    reducer=original_reducer,
                    axis=None,
                    mask=mask,
                    keepdims=keepdims,
                    behavior=behavior,
                )
```

File: src/awkward/_kernels.py (hunk 2 above)
Comment (ikrommyd) [reply src/awkward/_kernels.py:79]: Apart from this optimization not being used, a new kernel object is created every time you call a kernel and then called. So the attempt here was to move something from `__call__` to `__init__` but that does not offer speed up since it will be called exactly the same number of times. An actual optimization would be to do this in `__init__` but also cache kernel objects so that they don't have to be recreated the second time they are used. But that's a separate PR.

File: src/awkward/_reducers.py (hunk 3 above)
Comment (ikrommyd) [reply src/awkward/_reducers.py]: That doesn't really matter. The function can return `np.iinfo(dtype).max` which is an integer. So `np.generic` covers all numpy scalars.

File: src/awkward/_do.py (hunk 1 above)
Comment (ianna) [reply src/awkward/_do.py:324]: it calls a generic reducer, not AxisNoneReducer

Comment (ikrommyd) [reply src/awkward/_do.py:324]: `original_reducer` will be `Max` `Min` `Sum` etc. When called recursively with `axis=None`, it will end up in the `if axis is None` branch which will check if there is a specialization and get it.

File: src/awkward/_reducers.py (hunk 3 above)
Comment (ianna) [reply src/awkward/_reducers.py]: ok, thanks

File: src/awkward/_do.py (hunk 1 above)
Comment (ikrommyd) [reply src/awkward/_do.py:324]: `axis=0` and `axis=-1` hit the breakpoint when only one axis is available as expected exactly due to this recursion.
```diff
diff --git a/src/awkward/_reducers.py b/src/awkward/_reducers.py
index c7a991c8..3583bc43 100644
--- a/src/awkward/_reducers.py
+++ b/src/awkward/_reducers.py
@@ -1001,5 +1001,5 @@ class AxisNoneMax(Max):
                 result_scalar = nplike.maximum(result_scalar, initial_val)
 
         result_array = nplike.reshape(nplike.asarray(result_scalar), (1,))
-
+        breakpoint()
         return ak.contents.NumpyArray(result_array, backend=array.backend)
```
```py
In [1]: import awkward as ak
a
In [2]: array = ak.Array([1,2,3])

In [3]: ak.max(array, axis=0)
> /Users/iason/Dropbox/work/pyhep_dev/awkward/src/awkward/_reducers.py(1004)apply()
-> breakpoint()
(Pdb) c
Out[3]: np.int64(3)

In [4]: ak.max(array, axis=-1)
> /Users/iason/Dropbox/work/pyhep_dev/awkward/src/awkward/_reducers.py(1004)apply()
-> breakpoint()
(Pdb) c
Out[4]: np.int64(3)

In [5]: array = ak.Array([[1,2,3],[4,5]])

In [6]: ak.max(array, axis=None)
> /Users/iason/Dropbox/work/pyhep_dev/awkward/src/awkward/_reducers.py(1004)apply()
-> breakpoint()
(Pdb) c
Out[6]: np.int64(5)
```

**Comment by ikrommyd** [comment]
I can keep the jax code in and also add the an axis none min reducer for jax too, but I don't understand how it has anything to do with the blueprint. jax backend is slow, unjittable, and deprecated. How does this performance optimization help? Reductions work for jax, they just do not use this optimization. Also with the previous code, they were not being used for jax anyways because they jax reducers do not implement the `axis_none_reducer` method.

**Comment by ianna** [comment]
> I can keep the jax code in and also add the an axis none min reducer for jax too, but I don't understand how it has anything to do with the blueprint. jax backend is slow, unjittable, and deprecated. How does this performance optimization help? Reductions work for jax, they just do not use this optimization. Also with the previous code, they were not being used for jax anyways because they jax reducers do not implement the `axis_none_reducer` method.

Thanks! I will be taking about JAX and want to cross the tees — just to double check performance statements we made in the past.

File: src/awkward/_kernels.py (hunk 2 above)
Comment (ianna) [reply src/awkward/_kernels.py:79]: We do cache kernels

**Comment by ianna** [comment]
> > I can keep the jax code in and also add the an axis none min reducer for jax too, but I don't understand how it has anything to do with the blueprint. jax backend is slow, unjittable, and deprecated. How does this performance optimization help? Reductions work for jax, they just do not use this optimization. Also with the previous code, they were not being used for jax anyways because they jax reducers do not implement the `axis_none_reducer` method.
> 
> Thanks! I will be taking about JAX and want to cross the tees — just to double check performance statements we made in the past.

Oh, missed that part about implementing extra reducers for JAX — I don’t think it’s necessary at this point. Thanks

File: src/awkward/_kernels.py (hunk 2 above)
Comment (ikrommyd) [reply src/awkward/_kernels.py:79]: where?

**Comment by ikrommyd** [comment]
I'm not following perfectly. In the last PR, you attempted adding a fast-path axis none specialization for sum and max for jax? Do you want those in or not?

File: src/awkward/_kernels.py (hunk 2 above)
Comment (ikrommyd) [reply src/awkward/_kernels.py:79]: With a breakpoint here https://github.com/scikit-hep/awkward/blob/970f5d0e2057c5627fb9bf45bb0df16cdbe7786c/src/awkward/_backends/numpy.py#L32
I hit it every time I call a kernel. Where is the caching? A new `NumpyKernel` is created every time. And therefore `__init__` runs every time.
```
In [3]: array = ak.Array([1,2,3], backend="cpu")

In [4]: ak.min(array)
> /home/iason/Dropbox/work/pyhep_dev/awkward/src/awkward/_backends/numpy.py(32)__getitem__()
-> breakpoint()
(Pdb) c
Out[4]: np.int64(1)

In [5]: ak.min(array)
> /home/iason/Dropbox/work/pyhep_dev/awkward/src/awkward/_backends/numpy.py(32)__getitem__()
-> breakpoint()
(Pdb) c
Out[5]: np.int64(1)
```

**Comment by ianna** [comment]
> I'm not following perfectly. In the last PR, you attempted adding a fast-path axis none specialization for sum and max for jax? Do you want those in or not?

Yes, I want `sum` and `max` for JAX, but not `min`. Thanks.

**Comment by ikrommyd** [comment]
Then we'll have `min` too. Once the final release with the jax backend happens, it makes no sense for `max` to be much faster than `min`. I very highly want symmetry there and it's the same as max with changing max->min inside. All the jax changes are ready in this PR.

**Comment by ianna** [comment]
> Then we'll have `min` too. Once the final release with the jax backend happens, it makes no sense for `max` to be much faster than `min`. I very highly want symmetry there and it's the same as max with changing max->min inside. All the jax changes are ready in this PR.

I completely agree that `min` and `max` should eventually be symmetrical for the final JAX release. The current asymmetry is actually a deliberate part of my 'max specialization' testing; I'm using `min` as the control group to measure the exact latency reduction and throughput improvement we get from this specialized path compared to the generic one.

**Comment by ikrommyd** [comment]
> > Then we'll have `min` too. Once the final release with the jax backend happens, it makes no sense for `max` to be much faster than `min`. I very highly want symmetry there and it's the same as max with changing max->min inside. All the jax changes are ready in this PR.
> 
> I completely agree that `min` and `max` should eventually be symmetrical for the final JAX release. The current asymmetry is actually a deliberate part of my 'max specialization' testing; I'm using `min` as the control group to measure the exact latency reduction and throughput improvement we get from this specialized path compared to the generic one.

Ah I see, okay then I would suggest you do this locally or on another remote branch by just deleting the `AxisNoneMin` that I'm adding here for jax. I think we should treat main branch as "production" in general. Testing like this is very easy to do locally by just deleting a few lines. Developers may often run on "main". I run my analysis on "main" for example. That's why I think it should be kept clean like production. Like anyone should be able to install from main and not expect "weird" behaviors at any time, that would be the goal.

**Comment by ikrommyd** [comment]
@ianna removed the tests as those code paths were being hit anyways. Is this good to go now?

**Review by ianna — APPROVED** [review]
@ikrommyd - great! Thanks!
